### PR TITLE
Expand cursor object

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This WIP would use the cursor object instead of the vispy event for the mouse bindings. It contains some, but not all of the properties listed in #1048. An alternative would be to use an NamedTuple.

I'm not sure about this direction yet, but was exploring in the context of #1985 where we ultimately want to drop `layer.value`, `layer.coordinates`, `layer.displayed_coordinates`, `layer.position`.

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Comment by sofroniewn** [comment]
@jni this isn't ready for your review yet, but I'd love to get your feedback on it's main idea - which is swapping the `event` for a `cursor` object in the mouse function callback. I think we discussed this once and you didn't like the idea, which is fine, but I'm trying to think of an alternative now.

The issue is I'd like be able to drop `layer.coordinates`, `layer.displayed_coordinates`, and `layer.position` these shouldn't live on the layer itself, but then we need to be able to make these things available to our custom mouse functions somehow. The `Event` right now is just the `vispy` event, which we could wrap ourselves anyway to future proof that API, if we don't want to pass the `Cursor`.

Anyway I see this as an important step on the journey to removing slicing from the layers.

**Comment by jni** [comment]
@sofroniewn doesn't the event contain the x/y canvas coordinates? Can that be converted "easily" to world/data coordinates?

**Comment by jni** [comment]
One litmus test for the direction is whether this makes it easier or harder to have a "useful" cursor object in the 3D viewer. We had a nice discussion about 3D interactivity at the eurasia meeting yesterday. I'll copy @haesleinhuepf's pitch here:

> - Robert: Crop-plugin. Cool to limit computation on a region. However, some drag&drop interactivity for specifying the bounding-box would be cool and I'd love to hear hints/tips/tricks if something similar has been implemented already...
> 

@alisterburt mentioned/demoed dtmslice as a model for this sort of interactivity:

> - - Alister's MATLAB demo software: https://wiki.dynamo.biozentrum.unibas.ch/w/index.php/Downloads
> https://wiki.dynamo.biozentrum.unibas.ch/w/index.php/Dtmslice

**Comment by jni** [comment]
@sofroniewn could you elaborate on the plan here? Looking at the code I'm seeing a little bit better but I still don't understand how this helps to remove layer.coordinates... The cursor now has the coordinates, 👍, but is that needed? Can they not be computed dynamically from the event coordinates?

**Comment by sofroniewn** [comment]
> @sofroniewn doesn't the event contain the x/y canvas coordinates? Can that be converted "easily" to world/data coordinates?

There is `event.pos` which contains the x/y canvas info - these need to be translated into world coordinates, and this transform must happen on the qt_viewer right now, it can't happen on the layer as the layer doesn't know about the canvas position. they layer itself then must do the world to data conversion.

This is what we're pulling off the `Event` right now - see https://github.com/napari/napari/pull/1987/files#diff-1807f13d5d575e770c4b24ce95223949ab74d5492ec5eb1a4f71adfcf2e2531cR638

```python
            self.viewer.canvas_position = tuple(event.pos)
            self.viewer.cursor.is_dragging = bool(event.is_dragging)
            self.viewer.cursor.type = event.type
            if hasattr(event.native, 'inverted'):
                self.viewer.cursor.inverted = bool(event.native.inverted())
            else:
                self.viewer.cursor.inverted = False
            self.viewer.cursor.delta = tuple(event.delta)
            self.viewer.cursor.modifiers = tuple(event.modifiers)
        self.viewer.cursor.position = self._map_canvas2world(list(event.pos))
```

Critically the line is the `self._map_canvas2world(list(event.pos))` conversion, which needs to happen before this event get's passed to the layer or viewer.

The full mouse event API is here https://vispy.org/api/vispy.app.canvas.html?highlight=mouse%20event#vispy.app.canvas.MouseEvent

I see three options:

1) We swap the vispy mouse event for the expanded cursor object, which critically contains `cursor.position` the position in world coordinates. What's nice about this is we reuse our cursor object which is clearly documented and has everything needed.

2) We swap the vispy mouse event for our own mouse event, which would now critically contain `event.position` the position in world coordinates. What's nice about this is we have our own mouse event object which has a simpler API then the full vispy one and which we now will always work and be supported even if we moved to a webgl canvas or there were changes on the vispy end.

3) We add `event.position` to the current vispy event, which critically is the position in world coordinates. The downside of this is that it will be hard to document this new event, we can't point people to the vispy documentation as it will be missing this key attribute, we can't really type this object as a vispy object as it isn't anymore, and we'll need to create our own docs which include the info from the vispy docs including things that we've never tested / might not work as expect within their events.

I prefer either options (1) or (2) and dislike option (3). We might want to ask @Carreau his opinion here too, I know he's interested in adding touch events support through vispy too.

**Comment by jni** [comment]
A short summary of option 4 brought up on [Zulip](https://napari.zulipchat.com/#narrow/stream/270578-reviews/topic/.231987/near/224282993):

> @jni: I'm wondering if you could take all those transforms and make `cursor = napari.utils.cursor_from_mouse_event(viewer, event)`. It might not be any clearer, but it prevents fiddling with the event itself and it makes mouse events behave like mouse events.
>
> @sofroniewn: I could make CursorEvent from Event which just has what we want ...

To me that seems like the cleanest right now. But it would be great to get further input.

**Comment by Carreau** [comment]
I've rebased and expanded my work on the touch event on top of this here:  
https://github.com/Carreau/napari/pull/new/touch_and_mouse

VisPy support for passing touch event around is not that great; so one thing we might want to be careful about is what "Cursor" object may or may not have when not a mouse.

**Comment by sofroniewn** [comment]
> VisPy support for passing touch event around is not that great; so one thing we might want to be careful about is what "Cursor" object may or may not have when not a mouse.

Hmm yeah, I wonder if instead of expanding the cursor object like we've done here we add a napari `MouseEvent` and napari `TouchEvent` that are their own objects and customized for napari's needs as @jni suggested above. How would you feel about that @Carreau?

File: napari/layers/points/_points_mouse_bindings.py [hunk 1]
```diff
@@ -83,7 +83,7 @@ def select(layer, event):
     layer._set_highlight(force=True)
 
 
-def add(layer, event):
+def add(layer, cursor):
     """Add a new point at the clicked position."""
     # on press
     dragged = False
```
Comment (Carreau) [inline napari/layers/points/_points_mouse_bindings.py:89]: s/event/cursor/g line 93.

**Comment by sofroniewn** [comment]
This approach is being superseded by #2443

**Pull request closed without merging**